bson.c

```diff
@@ -849,25 +849,43 @@ char* bson_to_zval(char *buf, HashTable *result, mongo_bson_conversion_options *
 	 * bytes to that point.
 	 *
 	 * We lose buf's position as we iterate, so we need buf_start to save it. */
-	char *buf_start = buf, *buf_end;
+	const char *buf_start = buf, *buf_end;
 	unsigned char type;
 
 	if (buf == 0) {
 		return 0;
 	}
 
-	buf_end = buf + MONGO_32(*((int*)buf));
+	if (buf_len < 5) {
+		zend_throw_exception_ex(mongo_ce_CursorException, 38 TSRMLS_CC, "BSON buffer is %u bytes, but should be >= 5 (i.e. empty document)", buf_len);
+		return 0;
+	}
+
+	if (buf_len < (size_t) MONGO_32(*((int32_t*)buf))) {
+		zend_throw_exception_ex(mongo_ce_CursorException, 39 TSRMLS_CC, "BSON buffer is %u bytes, but document length is %u", buf_len, MONGO_32(*((int32_t*)buf)));
+		return 0;
```

Comment:
I'm very confused by this check.

when we hit BSON_ARRAY, it is literally the same value being compared, e.g. not achieving anything.

And what happens when the msg_header.length is reported correctly, but the BSON_ARRAY pretends to be 4gb?
All the first check does is to verify buffer underrun - while the second is a noop